src/freetype/ft-ogl.adb

```diff
@@ -0,0 +1,285 @@
+         if not GL.Objects.Textures.Is_Texture  (Char_Texture.Raw_Id) then
+            Put_Line ("FT.OGL.Render_Text, aTexture is invalid.");
+         end if;
+
+         GL.Objects.Textures.Set_Active_Unit (0);
+         Texture_2D.Bind (Char_Texture);
+         GL.Uniforms.Set_Int (Texture_ID, 0);
+         GL.Uniforms.Set_Single (Colour_ID, Colour (R), Colour (G), Colour (B));
+         GL.Uniforms.Set_Single (Projection_Matrix_ID, Projection_Matrix);
+
+         GL.Attributes.Enable_Vertex_Attrib_Array (0);
+         Array_Buffer.Bind (Vertex_Buffer);
+         GL.Attributes.Set_Vertex_Attrib_Pointer (Index  => 0, Count  => Num_Components,
+                                                  Kind   => GL.Types.Single_Type,
+                                                  Stride => Stride, Offset => 0);
+
+         GL.Objects.Vertex_Arrays.Draw_Arrays (Triangles, 0, Num_Vertices);
+         GL.Attributes.Disable_Vertex_Attrib_Array (0);
+         --  Bitshift by 6 to get value in pixels (2^6 = 64
+         --  (divide amount of 1/64th pixels by 64 to get amount of pixels))
+         X_Orig := X_Orig + Single (Advance_X (Char_Data)) / 64.0 * Scale;
+      end loop;
+      GL.Toggles.Disable (GL.Toggles.Blend);
+   end Render_Text;
+
+   --  ------------------------------------------------------------------------
+
+   procedure Set_Char_Data (Char_Data : in out Character_Record;
+                            Width     : GL.Types.Int; Height : GL.Types.Int;
+                            Left      : GL.Types.Int; Top    : GL.Types.Int;
+                            Advance_X : GL.Types.Int) is
+   begin
+      Char_Data.Width := Width;
+      Char_Data.Rows := Height;
+      Char_Data.Left := Left;
+      Char_Data.Top := Top;
+      Char_Data.Advance_X := Advance_X;
+   end Set_Char_Data;
+
+   --  -------------------------------------------------------------------------
+
+   procedure Setup_Character_Textures
+     (Face_Ptr  : FT.Faces.Face_Reference) is
+      use GL.Objects.Buffers;
+      use GL.Types;
+      Glyph_Slot     : constant Glyph_Slot_Reference := Face_Ptr.Glyph_Slot;
+      Width          : GL.Types.Size;
+      Height         : GL.Types.Size;
+      X_Offset       : constant GL.Types.Int := 0;
+      Y_Offset       : constant GL.Types.Int := 0;
+      Char_Data      : Character_Record;
+   begin
+      --  Blending allows a fragment colour's alpha value to control the resulting
+      --  colour which will be transparent for all the glyph's background colours and
+      --  non-transparent for the actual character pixels.
+      GL.Toggles.Enable (GL.Toggles.Blend);
+      GL.Blending.Set_Blend_Func (GL.Blending.Src_Alpha,
+                                  GL.Blending.One_Minus_Src_Alpha);
+
+      Vertex_Array.Initialize_Id;
+      Vertex_Array.Bind;
+
+      Vertex_Buffer.Initialize_Id;
+      Array_Buffer.Bind (Vertex_Buffer);
+
+      for index in Extended_Ascii_Data'Range loop
```

Comment:
On the long run, we'll want to be able to print more than just ANSI characters. Therefore, I wonder if this approach of preloading all character images is wise. How about this alternative approach:

Provide a type `Renderable_Text`. This can be created with a `Text_Renderer` from a String and is used as input for `Render_Text`. When creating `Renderable_Text`, you check whether each character is loaded and if not, load it. `Renderable_Text` then only contains an index list pointing to the characters used, and when you render it, you can be sure that the characters it uses are properly loaded.

Reply:
Preloading all character images originates from the original C source of this example, the context of which I would like to preserve.
I suggest making an additional example incorporating all of your ideas which certainly should be incorporated into the final "production" version of OGL.
As the current OGL falls well short of the ideal version that you have described, I'm wondering if OGL should be moved out of the current `freetype` library and back into the example itself until it is improved to its most useful implementation?